### PR TITLE
[Example] Support more ops for onnx models

### PR DESCRIPTION
**Pull request by pinzhenx**
https://github.com/intel/webml-polyfill/issues/222

- Enable BatchNormalization, Add, Mul, Reshape, Constant and depthwise convolution
- Fuse batch norm into (regular/depthwise) convolution
- Support Mobilenetv2 in WASM

### CODE REVIEW

**Comment by pinzhenx** [comment]
Fix https://github.com/intel/webml-polyfill/issues/240

**Comment by huningxin** [comment]
@pinzhenx Thanks for the PR! I will try to find time to provide feedbacks this week.

At meanwhile, @GreyZzzzzzXh , could you please review the modification for webgl2 backend?

**Comment by pinzhenx** [comment]
- Enable fully-connected-like Gemm
- Support Resnet v1/v2

File: examples/squeezenet/SqueezeNet.js [hunk 1]
```diff
@@ -68,12 +76,17 @@ class SqueezeNet {
   _addTensorOperands() {
     const graph = this._onnxModel.graph;
     for (let i = 0; i < graph.input.length; ++i) {
+      // delete useless empty node in the graph
+      if (graph.input[i].type.tensorType.elemType === 7 && graph.input[i].type.tensorType.shape.dim.length === 0) {
```
Comment (huningxin) [inline examples/squeezenet/SqueezeNet.js]: What is `elemType` as `7`? Could you define a named constant or some constant defined in the onnx.js?

File: examples/squeezenet/SqueezeNet.js [hunk 2]
```diff
@@ -145,26 +163,38 @@ class SqueezeNet {
         }
         data = nhwcData;
       }
+
+      // shape tensor for reshape
+      if (initializer.dims.length === 1 && initializer.int64Data.length) {
+        data = new Int32Array(initializer.int64Data);
+      }
```
Comment (huningxin) [inline examples/squeezenet/SqueezeNet.js]: would we show some warnings for other tensor types or shapes?

File: examples/squeezenet/SqueezeNet.js [hunk 3]
```diff
@@ -214,12 +251,12 @@ class SqueezeNet {
...
           const paddingWidthBegin = pads.ints[1];
           const paddingHeightEnd = pads.ints[2];
-          const paddingWidthEnd = pads.ints[3]
+          const paddingWidthEnd = pads.ints[3];
           inputs.push(this._addScalarInt32(paddingWidthBegin));
           inputs.push(this._addScalarInt32(paddingWidthEnd));
           inputs.push(this._addScalarInt32(paddingHeightBegin));
           inputs.push(this._addScalarInt32(paddingHeightEnd));
-
+
```
Comment (huningxin) [inline examples/squeezenet/SqueezeNet.js]: please remove the heading white spaces.

File: examples/squeezenet/utils.js [hunk 4]
```diff
@@ -56,6 +62,8 @@ class Utils {
     let start = performance.now();
     let result = await this.model.compute(this.inputTensor, this.outputTensor);
     let elapsed = performance.now() - start;
+    if (this.postOptions.softmax)
+      this.outputTensor = this.softmax(this.outputTensor);
```
Comment (huningxin) [inline examples/squeezenet/utils.js]: are we able to insert a softmax layer instead of computing it in JS?

**Comment by huningxin** [comment]
@pinzhenx , good work, thanks!

I've dropped some comments, please take a look.

File: examples/squeezenet/SqueezeNet.js (hunk 2 above)
Comment (pinzhenx) [reply examples/squeezenet/SqueezeNet.js]: I will add some warnings to type casting like this.

File: examples/squeezenet/SqueezeNet.js (hunk 1 above)
Comment (pinzhenx) [reply examples/squeezenet/SqueezeNet.js]: `7` stands for `onnx.TensorProto.DataType.INT64`. It's a dummy node in the onnx graph and this is just a workaround for supporting onnx mobilenet. Later, I will try to remove these lines.

**Comment by pinzhenx** [comment]
Should we add `getOperandValue()` to our API corresponding to setOperandValue()?

I have to obtain some processed initializer data in fusing ops. Now I directly access the `_model._operands[]` to get values, but I recently noticed that we can't access that array in WebML. So I'm wondering if it's necessary to add a new function to API, since I could alternatively maintain a list of id:initializer on my own.

**Comment by huningxin** [comment]
> Should we add `getOperandValue()` to our API corresponding to `setOperandValue()`?

That's a good question. But we don't have `getOperandValue()` so far.

> Now I directly access the `_model._operands[]` to get values, but I recently noticed that we can't access that array in WebML.
 
Please avoid to do that. In current `WebML`, they are not accessible.

Please maintain the data as your own at current stage.

File: examples/squeezenet/SqueezeNet.js (hunk 1 above)
Comment (huningxin) [reply examples/squeezenet/SqueezeNet.js]: If you want to keep this line, please use `onnx.TensorProto.DataType.INT64`, then it will be fine.

File: examples/squeezenet/utils.js (hunk 4 above)
Comment (pinzhenx) [reply examples/squeezenet/utils.js]: Absolutely. Now I implemented softmax as an extra layer. (and I will upload it later today)

I think it would be better to add an constructor option that allows user to decide if it should automatically append a softmax layer.
```javascript
constructor(onnxModel, backend, options = {}) { ... }
```

There are two benefits to doing so:
- More general to other applications other than image classification.  
- Easier to test. Though it's possible to add a softmax to the raw outputs of the test data set. But in some cases, such a layer will make lots of outputs to be zero so that we cannot compare two results effectively.

**Comment by pinzhenx** [comment]
- https://github.com/intel/webml-polyfill/pull/266/commits/a2b5cd826dd93cf3152acf4ee420b10bdf8a5720 adds support for Unsqueeze operator
- https://github.com/intel/webml-polyfill/pull/266/commits/a2b5cd826dd93cf3152acf4ee420b10bdf8a5720 adds support for multidirectional broadcasting in Add/Mul.
    - Support extending the tensor with fewer dimensions
- Support Inception V2, DenseNet in WASM
    - Only NNAPI support Add/Mul two operators with 'compatible' dimensions, i.e. broadcasting
    - Both models passed the test data sets provided by onnx zoo
    - Need to figure out image preprocessing for these two models https://github.com/onnx/models/issues/86

File: examples/squeezenet/SqueezeNet.js [hunk 5]
```diff
@@ -407,8 +717,27 @@ class SqueezeNet {
           console.warn(`    ${node.opType} is not supported.}`);
         }
       }
-      if (opCode)
+      if (typeof opCode !== 'undefined') {
         this._model.addOperation(opCode, inputs, outputs);
+      }
+      if (i === graph.node.length - 1 && node.opType !== 'Softmax') { // last node in graph?
```
Comment (huningxin) [inline examples/squeezenet/SqueezeNet.js]: Some object detection models might not have `softmax` at the last node. Do you have better way to add it? Perhaps create a configuration for specific model post-processing?

Comment (pinzhenx) [reply examples/squeezenet/SqueezeNet.js]: yes, i proposed a way to handle this.
https://github.com/intel/webml-polyfill/pull/266#discussion_r228756112

Comment (pinzhenx) [reply examples/squeezenet/SqueezeNet.js]: I have added an option to the constructor in https://github.com/intel/webml-polyfill/pull/266/commits/c8cc37c48d8441e62eaad7f43e3c0fb75f6810cd

If `modelOptions.softmax` is set to true, the importer will append a softmax layer called `softmax_appended` when the last layer is not a softmax. However, if this option is set to true but the model has already got a softmax layer, it won't do anything.

Comment (huningxin) [reply examples/squeezenet/SqueezeNet.js]: looks good.
